src/nsec3_fmt_plug.c

```diff
@@ -0,0 +1,396 @@
+	{ "$NSEC3$100$42424242$8fb38d13720815ed5b5fcefd973e0d7c3906ab02$example.com.", "mx" },
+	{ NULL }
+};
+
+
+static struct salt_t saved_salt;
+/* length of the saved label, without the length field */
+static int saved_key_length;
+static unsigned char saved_key[PLAINTEXT_LENGTH + 1];
+static unsigned char saved_wf_label[PLAINTEXT_LENGTH + 2];
+
+static SHA_CTX sha_ctx;
+static ARCH_WORD_32 crypt_out[5];
+
+static void convert_label_wf(void)
+{
+	int last_dot = saved_key_length - 1;
+	int i;
+	unsigned char *out = saved_wf_label;
+	if (saved_key_length == 0)
+		return;
+	++out;
+	for (i = last_dot ; i >= 0; ) {
+		if (saved_key[i] == '.') {
+			out[i] = (unsigned char)(last_dot - i);
+			last_dot = --i;
+		} else {
+			out[i] = tolower(saved_key[i]);
+			--i;
+		}
+	}
+	*(--out) = (unsigned char)(last_dot - i);
+}
+
+static size_t parse_zone(char *zone, unsigned char *zone_wf_out)
+{
+	char *lbl_end, *lbl_start;
+	unsigned int lbl_len;
+	unsigned int index = 0;
+	unsigned int zone_len = strlen(zone);
+
+	/* TODO: unvis */
+	if (zone_len == 0) {
+		return 0;
+	} else if (zone_len > N3_MAX_ZONE_SIZE) {
+		return 0;
+	}
+
+	lbl_end = strchr(zone, '.');
+	lbl_start = zone;
+	while (lbl_end != NULL) {
+		lbl_len = lbl_end - lbl_start;
+		zone_wf_out[index] = (unsigned char) lbl_len;
+		if (lbl_len > 0) {
+			memcpy(&zone_wf_out[++index], lbl_start, lbl_len);
+		}
+		index += lbl_len;
+		lbl_start = lbl_end+1;
+		if (lbl_start - zone == zone_len) {
+			zone_wf_out[index] = 0;
+			break;
+		} else {
+			lbl_end = strchr(lbl_start, '.');
+		}
+	}
+	if (lbl_end == NULL)
+		return 0;
+	return index + 1;
+}
+
+
+/* format:
+ * $NSEC3$iter$salt$hash$zone
+ */
+
+static int valid(char *ciphertext, struct fmt_main *pFmt)
+{
+	char *p, *q;
+	int i;
+	unsigned char zone[N3_MAX_ZONE_SIZE];
+	unsigned int iter;
+	char salt[N3_MAX_SALT_SIZE + 1];
+	char hash[HASH_LENGTH * 2 + 1];
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH))
+		return 0;
+	p = ciphertext;
+	for (i = 0; i < 4; ++i) {
+		p = strchr(p, '$');
+		if (p == NULL || *(++p) == 0)
+			return 0;
+		switch (i) {
+		case 0:
+			continue;
+		case 1:
+			/* iterations */
+			iter = atoi(p);
+			if (iter < 0 || iter > UINT16_MAX)
+				return 0;
+			break;
+		case 2:
+			/*  salt */
+			q = p;
+			while (atoi16[ARCH_INDEX(*q)] != 0x7F)
+				++q;
+			strncpy(salt, p, q - p);
+			salt[q - p] = 0;
+			if (!ishexlc(salt))
+				return 0;
+			if (*q != '$' || q-p > N3_MAX_SALT_SIZE*2 || (q-p) % 2)
+				return 0;
+			break;
+		case 3:
+			/* hash */
+			q = p;
+			while (atoi16[ARCH_INDEX(*q)] != 0x7F)
+				++q;
+			strncpy(hash, p, q - p);
+			hash[q - p] = 0;
+			if (!ishexlc(hash))
+				return 0;
+			if (*q != '$' || q-p > HASH_LENGTH*2)
+				return 0;
+			p = q+1;
+			break;
+		}
+	}
+	/* zone */
+	if (*p== 0)
+		return 0;
+	if (parse_zone(p, zone) == 0) {
+		return 0;
+	}
+	return 1;
+}
+
+static void *get_binary(char *ciphertext)
+{
+	static unsigned char out[BINARY_SIZE];
```

Comment:
You set BINARY_ALIGN to 4, but you will return unaligned if you use `unsigned char` here.